pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -589,7 +589,19 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
 	pipelineRunFacts.ResetSkippedCache()
 
 	// GetFinalTasks only returns tasks when a DAG is complete
-	nextRprts = append(nextRprts, pipelineRunFacts.GetFinalTasks()...)
```

Comment:
this is outside of the scope of this PR but i wonder if this function (`runNextSchedulableTask`) has too much responsibility. it seems to be in charge of:
- deciding what to execute next
- resolving references
- actually executing

those feel like at least 3 distinct responsibilities. i think if we get a chance it would be great to detangle this a bit.

Reply:
not to mention that it doesnt have unit tests :(

Reply:
😭 it does

I will work on refactoring it after this is merged, created issue #3610